### ISSUE TITLE
Implementation of T-SNE

### ISSUE THREAD

**Issue by charleswilmot** [google/jax#13186]
Hello all,

I could volunteer for implementing the `t-SNE` algorithm. 
[wikipedia](https://en.wikipedia.org/wiki/T-distributed_stochastic_neighbor_embedding)
[paper](https://www.jmlr.org/papers/volume9/vandermaaten08a/vandermaaten08a.pdf)

- Do you think that the algorithm should belongs to the JAX / scipy library?
- Can you think of an other repository where the `t-SNE` algorithm would fit better?

Best

**Comment by charleswilmot** [comment]
## [So here is a first version of my `t-SNE` implementation!](https://github.com/charleswilmot/jax_t_SNE)

```python
output_data = t_SNE(
    key,
    input_data,  # shape [B_0, ..., B_N, N_SAMPLES, N_FEATURES]
    desired_perplexity,
    reduction_dimension,
    update_func,
    termination_func,
    z_score=False
)   # shape [B_0, ..., B_N, N_SAMPLES, reduction_dimension]
```

I tried to stick to the functional philosophy of JAX.
I plan to add comments and typing to my code when I'll have time :-)

This implementation admits batched data.
- The axis `-2` is interpreted as the `n_samples` dimension.
- The axis `-1` is interpreted as the `n_features` dimension.
- All other dimensions are interpreted as `batch` dimensions.

To test the algorithm, you can just run the `t_SNE_examples.py` script.
You can chose between 2 examples:
- **MNIST_by_digit**   _(commented out)_ runs a batch of 9 t-SNE algorithms, for the digits 1 to 9
- **full_MNIST**   _(default)_ runs t-SNE on MNIST

_______________________________________________________________________________________________________

## Now I would also appreciate some advices :-)

The algorithm requires to compute distance matrices, which I do in this way:

```python
def squared_distance_matrix(data):
    return jnp.sum((
        data[..., None, :, :] -                       # [..., 1, N, D]
        data[..., None, :]                            # [..., N, 1, D]
    ) ** 2, axis=-1)                                  # [..., N, N]
```

However, this implementation is very memory intensive.
For example, for the MNIST dataset, this makes it hard to process more that 2048 samples.

**Would you have a proposal to compute the distance matrix in a more efficient manner ?**

Other implementation (ie. in scikit-learn for example) make use of K-nearest search algorithms, which I am not really willing to implement.

**Comment by charleswilmot** [comment]
Here you have it, now with comments, typing and pep8 formatting :-)
Cheers

**Comment by jakevdp** [comment]
Thanks! Looks good. I'm not entirely sure what to recomment regarding kNN searches. I don't know of any tree-based query that is compatible with JAX... perhaps scanning over batches would be the best approach?

**Comment by charleswilmot** [comment]
If by scanning you mean processing a batch in a more sequential manner, yes, that is what I would like to implement, however I am not sure what should be the preferred implementation.

Is there a nice way to do that with `jax`?

**Comment by jakevdp** [comment]
I don't know of any good example of this in JAX. You could probably do it using `lax.scan` to scan over batches of query points and then `lax.top_k` or `lax.approx_min_k` to extract the neighbors.

**Comment by charleswilmot** [comment]
I'll try to find some time to take a look at this after I defend my thesis on Friday

**Comment by jakevdp** [comment]
I'm going to close this as T-SNE is out-of-scope for JAX itself. But I'd encourage you to open-source the code elsewhere – it does seem like it would be useful to some people!
